### PR TITLE
Linked project spaces - specific tableau report permissions

### PR DESCRIPTION
**Pull request by AddisonDunn**
## Product Description

([Jira](https://dimagi-dev.atlassian.net/browse/USH-2257?atlOrigin=eyJpIjoiY2VjOWMxNTAzYTA5NDU0MmE2ZDM3ZTk4MWM3Y2E3YmEiLCJwIjoiaiJ9)) The delivery team has noticed that when doing a user roles push/pull for a linked project space the list of specific Tableau report permissions gets cleared, i.e.:

Screenshot 2023-04-05 at 4 00 28 PM

This occurred because we never added linked domains support for Tableau report permissions, which requires special attention.

This PR adds this support, which I hope is comprehensive.

## Technical Summary

The logic of how to handle a sync of Tableau report permissions is given in one of the code comments:
        # Enable permissions for visualization if the viz is enabled in the linked upstream viz
        # OR if the viz is enabled in the downstream project and doesn't have an upstream linked viz.

## Feature Flag

EMBEDDED_TALBEAU

### Safety story

Only affects domains with the EMBEDDED_TABLEAU FF and that sync user roles using the linked domains feature. If it does affect anything, it'll probably just cause an error or be pretty visible on the downstream domain's permissions list. Though this could _potentially_ have security concerns if something weren't caught early in this feature's adoption and web users synced without verifying afterwards.

### Automated test coverage

Added!

### QA Plan

None.

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/linked_domain/updates.py [hunk 1]
```diff
@@ -543,3 +553,15 @@ def _convert_reports_permissions(domain_link, master_results):
                 new_report_perms.append(get_ucr_class_name(linked_id))
 
         role_def['permissions']['view_report_list'] = new_report_perms
+
+
+def _get_new_tableau_report_permissions(downstream_domain_visualizations, upstream_permissions,
```
Comment (gherceg) [inline corehq/apps/linked_domain/updates.py:558]: Given that this is the core of this PR, I'd love to see tests around this method. The tests would also help document the expected behavior, and confirm that the issue with downstream permissions being undesirably overridden would not occur.

File: corehq/apps/linked_domain/updates.py [hunk 2]
```diff
@@ -543,3 +553,15 @@ def _convert_reports_permissions(domain_link, master_results):
...
+                                        original_downstream_permissions):
+    new_downstream_view_tableau_list = []
+    for viz in downstream_domain_visualizations:
+        # Enable permissions for visualization if the viz is enabled in the linked upstream viz
+        # OR if the viz is enabled in the downstream project and doesn't have an upstream linked viz.
+        if ((viz.upstream_id and viz.upstream_id in upstream_permissions.view_tableau_list)
+        or (not viz.upstream_id and str(viz.id) in original_downstream_permissions.view_tableau_list)):
```
Comment (gherceg) [inline corehq/apps/linked_domain/updates.py:565]: nit: I sometimes find it more readable to separate these larger conditionals into their own variables. The new variable names offer an opportunity to explain what that condition represents, and potentially remove the need for comments. Don't feel like you have to change this or feel free to change this as you like, but just as an example:
```suggestion
        is_upstream_visualization = viz.upstream_id and viz.upstream_id in upstream_permissions.view_tableau_list
        is_created_by_downstream = not viz.upstream_id and str(viz.id) in original_downstream_permissions.view_tableau_list
        if is_upstream_visualization or is_created_by_downstream:
```

**Review by gherceg — COMMENTED** [review]
Left a few comments. I do think tests for `_get_new_tableau_report_permissions` should be implemented as I don't think that will be a big lift, and ensures the logic introduced in this PR does what you want it to.

I'm also curious if you've talked to Matt about his work on destructive pushes within ERM. I don't think this will conflict with that work at all, but imagine that once that work is complete, the approach here could be updated to be consistent with that.

**Review by gherceg — APPROVED** [review]
Looks good to me and acknowledging your point about the existing test created here 👍🏻

**Review by esoergel — APPROVED** [review]

**Pull request merged**